Index: php/bbslib.h
===================================================================
--- php/bbslib.h	(revision 10670)
+++ php/bbslib.h	(working copy)
@@ -32,6 +32,7 @@
 int is_BM(const struct boardheader *board,const struct userec *user);
 char *unix_string(char *str);
 void output_ansi_html(char *buf, size_t buflen, buffered_output_t *output,char* attachlink, int is_tex, char* preview_attach_dir);
+void output_ansi_html_noatt(char *buf, size_t buflen, buffered_output_t *output);
 
 struct WWW_GUEST_S* www_get_guest_entry(int idx);
 
Index: php/phpbbslib.c
===================================================================
--- php/phpbbslib.c	(revision 10670)
+++ php/phpbbslib.c	(working copy)
@@ -307,6 +307,8 @@
     REGISTER_LONG_CONSTANT("BBS_PERM_ADMIN", PERM_ADMIN, CONST_CS | CONST_PERSISTENT);
     REGISTER_LONG_CONSTANT("BBS_PERM_WELCOME", PERM_WELCOME, CONST_CS | CONST_PERSISTENT);
     REGISTER_LONG_CONSTANT("BBS_PERM_ANNOUNCE", PERM_ANNOUNCE, CONST_CS | CONST_PERSISTENT);
+    REGISTER_LONG_CONSTANT("BBS_PERM_OBOARDS", PERM_OBOARDS, CONST_CS | CONST_PERSISTENT);
+    REGISTER_LONG_CONSTANT("BBS_PERM_CHAT", PERM_CHAT, CONST_CS | CONST_PERSISTENT);
     REGISTER_LONG_CONSTANT("BBS_BOARD_ATTACH", BOARD_ATTACH, CONST_CS | CONST_PERSISTENT);
     REGISTER_LONG_CONSTANT("BBS_BOARD_ANNONY", BOARD_ANNONY, CONST_CS | CONST_PERSISTENT);
     REGISTER_LONG_CONSTANT("BBS_BOARD_JUNK", BOARD_JUNK, CONST_CS | CONST_PERSISTENT);
@@ -318,6 +320,8 @@
     REGISTER_LONG_CONSTANT("BBS_BOARD_POSTSTAT", BOARD_POSTSTAT, CONST_CS | CONST_PERSISTENT);
     REGISTER_LONG_CONSTANT("BBS_BOARD_EMAILPOST", BOARD_EMAILPOST, CONST_CS | CONST_PERSISTENT);
     REGISTER_LONG_CONSTANT("BBS_BOARD_NOREPLY", BOARD_NOREPLY, CONST_CS | CONST_PERSISTENT);
+    REGISTER_LONG_CONSTANT("BBS_BOARD_READONLY", BOARD_READONLY, CONST_CS | CONST_PERSISTENT);
+    REGISTER_LONG_CONSTANT("BBS_BOARD_TMP_POST", BOARD_TMP_POST, CONST_CS | CONST_PERSISTENT);
     REGISTER_LONG_CONSTANT("BBS_MAXATTACHMENTCOUNT", MAXATTACHMENTCOUNT, CONST_CS | CONST_PERSISTENT);
     REGISTER_LONG_CONSTANT("BBS_MAXATTACHMENTSIZE", MAXATTACHMENTSIZE, CONST_CS | CONST_PERSISTENT);
 
Index: php/phpbbs_user.c
===================================================================
--- php/phpbbs_user.c	(revision 10670)
+++ php/phpbbs_user.c	(working copy)
@@ -31,6 +31,10 @@
     add_assoc_long(array, "signum", ud.signum);
     add_assoc_long(array, "userdefine0", user->userdefine[0]);
     add_assoc_long(array, "userdefine1", user->userdefine[1]);
+#ifdef NFORUM
+    add_assoc_long(array, "uid", num);
+    add_assoc_long(array, "mailbox_prop", getSession()->currentuinfo->mailbox_prop);
+#endif
 
 #ifdef HAVE_BIRTHDAY
     add_assoc_long(array,"gender",ud.gender);
Index: php/phpbbs_file.c
===================================================================
--- php/phpbbs_file.c	(revision 10670)
+++ php/phpbbs_file.c	(working copy)
@@ -451,6 +451,47 @@
     RETURN_STRINGL(get_output_buffer(), get_output_buffer_len(),1);
 }
 
+#ifdef NFORUM
+/**
+ * output without attachment modify by xw 20090919
+ */
+PHP_FUNCTION(bbs_printansifile_noatt)
+{
+    char *filename;
+    int filename_len;
+    long is_preview;
+    char *ptr;
+    off_t ptrlen, mmap_ptrlen;
+    const int outbuf_len = 4096;
+    buffered_output_t *out;
+
+    if (ZEND_NUM_ARGS() == 1) {
+        if (zend_parse_parameters(1 TSRMLS_CC, "s", &filename, &filename_len) != SUCCESS) {
+            WRONG_PARAM_COUNT;
+        }
+        is_preview=0;
+    } else if (ZEND_NUM_ARGS() == 2) {
+        if (zend_parse_parameters(2 TSRMLS_CC, "sl", &filename, &filename_len, &is_preview) != SUCCESS) {
+            WRONG_PARAM_COUNT;
+        }
+    }
+    if (!is_preview) {
+        if (safe_mmapfile(filename, O_RDONLY, PROT_READ, MAP_SHARED, &ptr, &mmap_ptrlen, NULL) == 0)
+            RETURN_LONG(-1);
+        ptrlen = mmap_ptrlen;
+    }
+    if ((out = alloc_output(outbuf_len)) == NULL) {
+        if (!is_preview) end_mmapfile(ptr, mmap_ptrlen, -1);
+        RETURN_LONG(2);
+    }
+    override_default_write(out, new_write);
+    output_ansi_html_noatt(ptr, ptrlen, out);
+    free_output(out);
+    if (!is_preview) end_mmapfile(ptr, mmap_ptrlen, -1);
+    RETURN_STRINGL(get_output_buffer(), get_output_buffer_len(),1);
+}
+#endif
+
 PHP_FUNCTION(bbs_print_article)
 {
     char *filename;
Index: php/phpbbs_file.h
===================================================================
--- php/phpbbs_file.h	(revision 10670)
+++ php/phpbbs_file.h	(working copy)
@@ -7,6 +7,7 @@
 PHP_FUNCTION(bbs2_readfile_text);
 PHP_FUNCTION(bbs_file_output_attachment);
 PHP_FUNCTION(bbs_printansifile);
+PHP_FUNCTION(bbs_printansifile_noatt);
 PHP_FUNCTION(bbs_print_article);
 PHP_FUNCTION(bbs_print_article_js);
 PHP_FUNCTION(bbs_printoriginfile);
@@ -20,6 +21,7 @@
     PHP_FE(bbs2_readfile_text, NULL) \
     PHP_FE(bbs_file_output_attachment, NULL) \
     PHP_FE(bbs_printansifile, NULL) \
+    PHP_FE(bbs_printansifile_noatt, NULL) \
     PHP_FE(bbs_print_article, NULL) \
     PHP_FE(bbs_print_article_js, NULL) \
     PHP_FE(bbs_printoriginfile, NULL) \
Index: php/phpbbs_post.c
===================================================================
--- php/phpbbs_post.c	(revision 10670)
+++ php/phpbbs_post.c	(working copy)
@@ -278,6 +278,31 @@
     }
 }
 
+PHP_FUNCTION(bbs_file_attachment_list)
+{
+    struct ea_attach_info ai[MAXATTACHMENTCOUNT];
+    char* filename;
+    int filename_len;
+    int ret,fd;    
+    int ac = ZEND_NUM_ARGS();
+
+    if (ac != 1 || zend_parse_parameters(1 TSRMLS_CC, "s", &filename, &filename_len) == FAILURE) {
+        WRONG_PARAM_COUNT;
+    }    
+
+    fd = open(filename, O_RDONLY);
+    if (fd < 0) 
+        RETURN_ERROR(GENERAL);
+
+    ret = ea_locate(fd, ai); 
+    close(fd);
+
+    if (ret<0 || dump_attachment_info(return_value, ai)) {
+        RETURN_ERROR(GENERAL);
+    }    
+
+}
+
 PHP_FUNCTION(bbs_attachment_list)
 {
     struct ea_attach_info ai[MAXATTACHMENTCOUNT];
@@ -1121,7 +1146,28 @@
 #endif
 }
 
+/** 
+ * bbs_postfile(filename, boardname, title)
+ * post via deliver
+ */
+PHP_FUNCTION(bbs_postfile)
+{
+    char *filename ,*boardname ,*filetitle;
+    int  filename_len ,boardname_len ,filetitle_len ,postid;
+        
+    int ac = ZEND_NUM_ARGS();
+    
+    if (ac != 3 || zend_parse_parameters(ZEND_NUM_ARGS()TSRMLS_CC, "sss" , &filename, &filename_len, &boardname, &boardname_len, &filetitle, &filetitle_len) == FAILURE)
+        WRONG_PARAM_COUNT;
 
+    init_all();
+    postid=post_file(NULL, "", filename, boardname, filetitle, 0, 1, getSession());
+    if ( postid< 0) {
+        RETURN_LONG(-1);
+    }
+    RETURN_LONG(postid);
+}
+
 PHP_FUNCTION(bbs_post_file_alt)
 {
     int ac, fname_len, userid_len, title_len, to_board_len, from_board_len, ret;
Index: php/phpbbs_board.c
===================================================================
--- php/phpbbs_board.c	(revision 10670)
+++ php/phpbbs_board.c	(working copy)
@@ -19,8 +19,26 @@
     add_assoc_long(array, "TOTAL", bstatus->total);
 }
 
+#ifdef NFORUM
+static void assign_board_nforum(zval * array, const struct boardheader *board, const struct BoardStatus* bstatus, int num)
+{
+    add_assoc_long(array, "BID", num);
+    add_assoc_string(array, "NAME", (char*)board->filename, 1);
+    add_assoc_string(array, "BM", (char*)board->BM, 1);
+    add_assoc_long(array, "FLAG", board->flag);
+    add_assoc_string(array, "DESC", (char*)board->title + 13, 1);
+    add_assoc_stringl(array, "CLASS", (char*)board->title + 1, 6, 1);
+    add_assoc_stringl(array, "SECNUM", (char*)board->title, 1, 1);
+    add_assoc_long(array, "LEVEL", board->level);
+    add_assoc_long(array, "CURRENTUSERS", bstatus->currentusers);
+    add_assoc_long(array, "GROUP", board->group);
+    add_assoc_long(array, "LASTPOST", bstatus->lastpost);
+    add_assoc_long(array, "ARTCNT", bstatus->total);
+    add_assoc_long(array, "UNREAD", 0);
+    add_assoc_long(array, "NPOS", getbid(board->filename,&board)-1);
+}
+#endif 
 
-
 char *brd_col_names[] = {
     "NAME",
     "DESC",
@@ -79,6 +97,26 @@
     }
 }
 
+#ifdef NFORUM
+static void assign_board_zval_nforum(zval * array, struct newpostdata *brd,const struct boardheader *board, const struct BoardStatus* bstatus,int num)
+{
+    add_assoc_long(array, "BID", num);
+    add_assoc_string(array, "NAME", (char*)board->filename, 1);
+    add_assoc_string(array, "BM", (char*)board->BM, 1);
+    add_assoc_long(array, "FLAG", board->flag);
+    add_assoc_string(array, "DESC", (char*)board->title + 13, 1);
+    add_assoc_stringl(array, "CLASS", (char*)board->title + 1, 6, 1);
+    add_assoc_stringl(array, "SECNUM", (char*)board->title, 1, 1);
+    add_assoc_long(array, "LEVEL", board->level);
+    add_assoc_long(array, "GROUP", board->group);
+    add_assoc_long(array, "CURRENTUSERS", bstatus->currentusers);
+    add_assoc_long(array, "LASTPOST", bstatus->lastpost);
+    add_assoc_long(array, "ARTCNT", bstatus->total);
+    add_assoc_long(array, "UNREAD", brd->unread);
+    add_assoc_long(array, "NPOS", brd->pos);
+}
+#endif
+
 static void assign_board_zval(zval * array, struct newpostdata *brd)
 {
     add_assoc_string(array, "NAME", (char *)brd->name, 1);
@@ -137,8 +175,26 @@
     add_assoc_long(array, "LASTPOST", 0);
 }
 
+#ifdef NFORUM
+static void assign_favdir_zval_nforum(zval * array,struct newpostdata *brd)
+{
+    add_assoc_long(array, "BID", -1);
+    add_assoc_string(array, "NAME", (char *)brd->name, 1);
+    add_assoc_string(array, "BM", (char *)brd->BM, 1);
+    add_assoc_long(array, "FLAG", -1L);
+    add_assoc_string(array, "DESC", (char *)brd->title, 1);
+    add_assoc_string(array, "CLASS", "", 1);
+    add_assoc_long(array, "SECNUM", -1);
+    add_assoc_long(array, "LEVEL", -1);
+    add_assoc_long(array, "GROUP", -1);
+    add_assoc_long(array, "CURRENTUSERS", -1);
+    add_assoc_long(array, "LASTPOST", -1);
+    add_assoc_long(array, "ARTCNT", -1);
+    add_assoc_long(array, "UNREAD", -1);
+    add_assoc_long(array, "NPOS", brd->tag);
+}
+#endif
 
-
 /* TODO: move this function into bbslib. */
 /* no_brc added by atppp 20040706 */
 static int check_newpost(struct newpostdata *ptr, bool no_brc)
@@ -173,9 +229,28 @@
     return 1;
 }
 
+#ifdef NFORUM
+PHP_FUNCTION(bbs_getboard_bid){
+    zval *array;
+    const struct boardheader *bh;
+    const struct BoardStatus *bs;
+    int b_num;
 
+    if (zend_parse_parameters(2 TSRMLS_CC, "la", &b_num, &array) != SUCCESS)
+        WRONG_PARAM_COUNT;
+    bh = getboard(b_num);
+    if(!bh)
+        RETURN_LONG(0);
+    bs = getbstatus(b_num);
+    if (array) {
+        if (array_init(array) != SUCCESS)
+            WRONG_PARAM_COUNT;
+        assign_board(array, bh, bs, b_num);
+    }
+    RETURN_LONG(b_num);
+}
+#endif
 
-
 PHP_FUNCTION(bbs_getboard)
 {
     zval *array;
@@ -1102,3 +1177,231 @@
     RETURN_LONG(deny_me(userid, bname));
 }
 
+#ifdef NFORUM
+
+PHP_FUNCTION(bbs_getboard_nforum)
+{
+    zval *array;
+    char *boardname;
+    int boardname_len;
+    const struct boardheader *bh; 
+    const struct BoardStatus *bs; 
+    int b_num;
+
+    if (ZEND_NUM_ARGS() == 1) { 
+        if (zend_parse_parameters(1 TSRMLS_CC, "s", &boardname, &boardname_len) != SUCCESS)
+            WRONG_PARAM_COUNT;
+        array = NULL;
+    } else {
+        if (ZEND_NUM_ARGS() == 2) { 
+            if (zend_parse_parameters(2 TSRMLS_CC, "sa", &boardname, &boardname_len, &array) != SUCCESS)
+                WRONG_PARAM_COUNT;
+        } else 
+            WRONG_PARAM_COUNT;
+    }    
+    if (boardname_len > BOARDNAMELEN)
+        boardname[BOARDNAMELEN] = 0; 
+    b_num = getbid(boardname, &bh);
+    if (b_num == 0)
+        RETURN_LONG(0);
+    bs = getbstatus(b_num);
+    if (array) {
+        if (array_init(array) != SUCCESS)
+            WRONG_PARAM_COUNT;
+        assign_board_nforum(array, bh, bs, b_num);
+    }    
+    RETURN_LONG(b_num);
+}
+
+PHP_FUNCTION(bbs_getboards_nforum)
+{
+    /*
+     * TODO: The name of "yank" must be changed, this name is totally
+     * shit, but I don't know which name is better this time.
+     */
+    char *prefix;
+    int plen;
+    long flag, group;
+    struct newpostdata *newpost_buffer;
+    struct newpostdata *ptr;
+    zval *element;
+    int i;
+    int j;
+    int ac = ZEND_NUM_ARGS();
+    int brdnum, yank, no_brc, all_boards;
+    int total;
+    const struct boardheader *bh; 
+    const struct BoardStatus *bs;
+    /*
+     * getting arguments
+     */
+    if (ac != 3 || zend_parse_parameters(3 TSRMLS_CC, "sll", &prefix, &plen, &group,&flag) == FAILURE) {
+        WRONG_PARAM_COUNT;
+    }
+
+    if (plen == 0) {
+        RETURN_FALSE;
+    }
+    if (getCurrentUser() == NULL) {
+        RETURN_FALSE;
+    }
+
+    total=get_boardcount();
+
+    yank = flag & 1;
+    no_brc = flag & 2;
+    all_boards = (flag & 4) && (group == 0);
+
+    brdnum = 0;
+    {
+        int n;
+        struct boardheader const *bptr;
+        const char** namelist;
+        int* indexlist;
+        time_t tnow;
+        int b_num;
+        tnow = time(0);
+        namelist=(const char**)emalloc(sizeof(char**)*(total));
+        if (namelist==NULL) {
+            RETURN_FALSE;
+        }
+        indexlist=(int*)emalloc(sizeof(int*)*(total));
+        if (indexlist==NULL) {
+            RETURN_FALSE;
+        }
+        for (n = 0; n < total; n++) {
+            bptr = getboard(n + 1);
+            if (!bptr)
+                continue;
+            if (*(bptr->filename)==0)
+                continue;
+            if (group == -2) {
+                if ((tnow - bptr->createtime) > 86400*30 || (bptr->flag & BOARD_GROUP))
+                    continue;
+            } else if (!all_boards && (bptr->group!=group))
+                continue;
+            if (!check_see_perm(getCurrentUser(),bptr)) {
+                continue;
+            }
+            if ((group==0)&&(strchr(prefix, bptr->title[0]) == NULL && prefix[0] != '*'))
+                continue;
+            /* if (yank || getSession()->zapbuf[n] != 0 || (bptr->level & PERM_NOZAP)) */ {
+                /*都要排序*/
+                for (i=0;i<brdnum;i++) {
+                    if (strcasecmp(namelist[i], bptr->filename)>0)
+                        break;
+                }
+                for (j=brdnum;j>i;j--) {
+                    namelist[j]=namelist[j-1];
+                    indexlist[j]=indexlist[j-1];
+                }
+                namelist[i]=bptr->filename;
+                indexlist[i]=n;
+                brdnum++;
+            }
+        }
+        newpost_buffer = (struct newpostdata*)emalloc(sizeof(struct newpostdata) * brdnum);
+        for (i=0;i<brdnum;i++) {
+            ptr=&(newpost_buffer[i]);
+            bptr = getboard(indexlist[i]+1);
+            ptr->dir = bptr->flag&BOARD_GROUP?1:0;
+            ptr->name = (char*)bptr->filename;
+            ptr->title = (char*)bptr->title;
+            ptr->BM = (char*)bptr->BM;
+            ptr->flag = bptr->flag | ((bptr->level & PERM_NOZAP) ? BOARD_NOZAPFLAG : 0);
+            ptr->pos = indexlist[i];
+            if (bptr->flag&BOARD_GROUP) {
+                ptr->total = bptr->board_data.group_total;
+            } else ptr->total=-1;
+            ptr->zap = (getSession()->zapbuf[indexlist[i]] == 0);
+            check_newpost(ptr, no_brc);
+        }
+
+        if (array_init(return_value) == FAILURE) {
+            RETURN_FALSE;
+        }
+        for (i=0;i<brdnum;i++) {
+            MAKE_STD_ZVAL(element);
+            array_init(element);
+            b_num = getbid(newpost_buffer[i].name, &bh);
+            bs = getbstatus(b_num);
+            assign_board_zval_nforum(element, &(newpost_buffer[i]),bh,bs,b_num);
+            zend_hash_index_update(Z_ARRVAL_P(return_value), i, (void *)&element, sizeof(zval*), NULL);
+        }
+        efree(newpost_buffer);
+        efree(namelist);
+        efree(indexlist);
+    }
+
+}
+
+PHP_FUNCTION(bbs_fav_boards_nforum)
+{
+    long select;
+    long mode;
+    struct newpostdata newpost_buffer[FAVBOARDNUM];
+    struct newpostdata *ptr;
+    zval *element;
+    int i;
+    int brdnum;
+    const struct boardheader *bh; 
+    const struct BoardStatus *bs;
+    int b_num;
+    /*
+     * getting arguments
+     */
+    if (zend_parse_parameters(ZEND_NUM_ARGS() TSRMLS_CC, "ll", &select, &mode) == FAILURE) {
+        WRONG_PARAM_COUNT;
+    }
+
+    /*
+     * loading boards
+     */
+    /*
+     * handle some global variables: getCurrentUser(), yank, brdnum,
+     * * nbrd.
+     */
+    /*
+     * NOTE: getCurrentUser() SHOULD had been set in funcs.php,
+     * * but we still check it.
+     */
+
+    if (mode==2) {
+        load_favboard(2, getSession());
+        if (select>=0 && select<favbrd_list_t)
+            SetFav(select, getSession());
+    } else if (mode==3) {
+        load_favboard(3, getSession());
+        if (select>=0 && select<favbrd_list_t)
+            SetFav(select, getSession());
+    }
+
+    if (getCurrentUser() == NULL) {
+        RETURN_FALSE;
+    }
+    brdnum = 0;
+
+    if ((brdnum = fav_loaddata(newpost_buffer, select, 1, FAVBOARDNUM, 1, NULL, getSession())) <= -1) {
+        RETURN_FALSE;
+    }
+
+    /** fill data in output array.*/
+    if (array_init(return_value) == FAILURE) {
+        RETURN_FALSE;
+    }
+    for (i=0;i<brdnum;i++) {
+        MAKE_STD_ZVAL(element);
+        array_init(element);
+        ptr = &newpost_buffer[i];
+        if (ptr->flag == 0xffffffff){ /* the item is a directory */
+            assign_favdir_zval_nforum(element, &(newpost_buffer[i]));
+        }
+        else{
+            b_num = getbid(newpost_buffer[i].name, &bh);
+            bs = getbstatus(b_num);
+            assign_board_zval_nforum(element, &(newpost_buffer[i]),bh,bs,b_num);
+        }
+        zend_hash_index_update(Z_ARRVAL_P(return_value), i, (void *)&element, sizeof(zval*), NULL);
+    }
+}
+#endif
Index: php/phpbbs_post.h
===================================================================
--- php/phpbbs_post.h	(revision 10670)
+++ php/phpbbs_post.h	(working copy)
@@ -11,6 +11,7 @@
 PHP_FUNCTION(bbs_attachment_add);
 PHP_FUNCTION(bbs_attachment_del);
 PHP_FUNCTION(bbs_attachment_list);
+PHP_FUNCTION(bbs_file_attachment_list);
 
 PHP_FUNCTION(bbs_postarticle);
 
@@ -23,6 +24,7 @@
 PHP_FUNCTION(bbs_docross);
 PHP_FUNCTION(bbs_docommend);
 PHP_FUNCTION(bbs_post_file_alt);
+PHP_FUNCTION(bbs_postfile);
 
 PHP_FUNCTION(bbs_brcaddread);
 PHP_FUNCTION(bbs_brcclear);
@@ -39,6 +41,7 @@
     PHP_FE(bbs_attachment_add,NULL) \
     PHP_FE(bbs_attachment_del,NULL) \
     PHP_FE(bbs_attachment_list,NULL) \
+    PHP_FE(bbs_file_attachment_list,NULL) \
     PHP_FE(bbs_postarticle,NULL) \
     PHP_FE(bbs_delpost,NULL) \
     PHP_FE(bbs_article_deny_modify,NULL) \
@@ -49,6 +52,7 @@
     PHP_FE(bbs_docross,NULL) \
     PHP_FE(bbs_docommend,NULL) \
     PHP_FE(bbs_post_file_alt,NULL) \
+    PHP_FE(bbs_postfile, NULL)\
     PHP_FE(bbs_brcaddread, NULL) \
     PHP_FE(bbs_brcclear, NULL) \
     PHP_FE(bbs2_brcdump, NULL) \
Index: php/phpbbs_board.h
===================================================================
--- php/phpbbs_board.h	(revision 10670)
+++ php/phpbbs_board.h	(working copy)
@@ -6,6 +6,9 @@
 PHP_FUNCTION(bbs_getboard);
 PHP_FUNCTION(bbs_safe_getboard);
 PHP_FUNCTION(bbs_getboards);
+PHP_FUNCTION(bbs_getboard_bid);
+PHP_FUNCTION(bbs_getboard_nforum);
+PHP_FUNCTION(bbs_getboards_nforum);
 
 PHP_FUNCTION(bbs_checkorigin);
 PHP_FUNCTION(bbs_checkmark);
@@ -24,6 +27,7 @@
 /* favboard operation. by caltary  */
 PHP_FUNCTION(bbs_load_favboard);
 PHP_FUNCTION(bbs_fav_boards);
+PHP_FUNCTION(bbs_fav_boards_nforum);
 PHP_FUNCTION(bbs_is_favboard);
 PHP_FUNCTION(bbs_add_favboarddir);
 PHP_FUNCTION(bbs_add_favboard);
@@ -58,7 +62,10 @@
     PHP_FE(bbs_get_father,NULL) \
     PHP_FE(bbs_get_dirname,NULL) \
     PHP_FE(bbs_del_favboarddir,NULL) \
-    PHP_FE(bbs_deny_me,NULL)
-
+    PHP_FE(bbs_deny_me,NULL)\
+    PHP_FE(bbs_getboard_bid,NULL)\
+    PHP_FE(bbs_getboard_nforum,NULL)\
+    PHP_FE(bbs_getboards_nforum,NULL)\
+    PHP_FE(bbs_fav_boards_nforum,NULL) 
 #endif //PHP_BBS_BOARD_H
 
Index: php/bbslib.c
===================================================================
--- php/bbslib.c	(revision 10670)
+++ php/bbslib.c	(working copy)
@@ -1120,7 +1120,184 @@
 
 }
 
+#ifdef NFORUM
+/**
+ * make buf with ansi without attachment
+ * see more in output_ansi_html
+ * modify by xw 20090919
+ */
 
+void output_ansi_html_noatt(char *buf, size_t buflen, buffered_output_t * output)
+{
+    unsigned int font_style = 0;
+    unsigned int ansi_state;
+    unsigned int ansi_val[STRLEN];
+    int ival = 0;
+    size_t i;
+    char *ansi_begin = 0;
+    char *ansi_end;
+    size_t article_len = buflen;
+	long attach_len;
+	char *attachptr, *attachfilename;
+
+    if (buf == NULL)
+        return;
+    STATE_ZERO(ansi_state);
+    bzero(ansi_val, sizeof(ansi_val));
+
+	//find the length of artilce without attachment
+	for (i = 0; i < buflen ; i++) {
+		if (((attachfilename = checkattach(buf + i, buflen - i, &attach_len, &attachptr)) != NULL)) {
+			article_len = attachfilename - buf - ATTACHMENT_SIZE;
+			break;
+		}
+	}
+
+    for (i = 0; i < article_len; i++) {
+        if (STATE_ISSET(ansi_state, STATE_NEW_LINE)) {
+            STATE_CLR(ansi_state, STATE_NEW_LINE);
+            if (i < (buflen - 1) && !STATE_ISSET(ansi_state,STATE_TEX_SET) && (buf[i] == ':' && buf[i + 1] == ' ')) {
+                STATE_SET(ansi_state, STATE_QUOTE_LINE);
+                if (STATE_ISSET(ansi_state, STATE_FONT_SET))
+                    BUFFERED_OUTPUT(output, "</font>", 7);
+                /*
+                 * set quoted line styles
+                 */
+                STYLE_SET(font_style, FONT_STYLE_QUOTE);
+                STYLE_SET_FG(font_style, FONT_COLOR_QUOTE);
+                STYLE_CLR_BG(font_style);
+                print_font_style(font_style, output);
+                BUFFERED_OUTPUT(output, &buf[i], 1);
+                STATE_SET(ansi_state, STATE_FONT_SET);
+                STATE_CLR(ansi_state, STATE_ESC_SET);
+                /*
+                 * clear ansi_val[] array
+                 */
+                bzero(ansi_val, sizeof(ansi_val));
+                ival = 0;
+                continue;
+            } else
+                STATE_CLR(ansi_state, STATE_QUOTE_LINE);
+        }
+        /*
+        * is_tex 情况下，\[upload 优先匹配 \[ 而不是 [upload
+        * is_tex 情况下应该还有一个问题是 *[\[ 等，不过暂时不管了 - atppp
+        */
+        if (i < (buflen - 1) && !STATE_ISSET(ansi_state,STATE_TEX_SET) && (buf[i] == 0x1b && buf[i + 1] == '[')) {
+            if (STATE_ISSET(ansi_state, STATE_ESC_SET)) {
+                /*
+                 *[*[ or *[13;24*[ */
+                size_t len;
+
+                ansi_end = &buf[i - 1];
+                len = ansi_end - ansi_begin + 1;
+                print_raw_ansi(ansi_begin, len, output);
+            }
+            STATE_SET(ansi_state, STATE_ESC_SET);
+            ansi_begin = &buf[i];
+            i++;                /* skip the next '[' character */
+        } else if (buf[i] == '\n') {
+            if (STATE_ISSET(ansi_state, STATE_ESC_SET)) {
+                /*
+                 *[\n or *[13;24\n */
+                size_t len;
+
+                ansi_end = &buf[i - 1];
+                len = ansi_end - ansi_begin + 1;
+                print_raw_ansi(ansi_begin, len, output);
+                STATE_CLR(ansi_state, STATE_ESC_SET);
+            }
+            if (STATE_ISSET(ansi_state, STATE_QUOTE_LINE)) {
+                /*
+                 * end of a quoted line
+                 */
+                BUFFERED_OUTPUT(output, "</font>", 7);
+                STYLE_CLR(font_style, FONT_STYLE_QUOTE);
+                STATE_CLR(ansi_state, STATE_FONT_SET);
+            }
+            if (!STATE_ISSET(ansi_state,STATE_TEX_SET)) {
+                BUFFERED_OUTPUT(output, " <br /> ", 8);
+            }
+            STATE_CLR(ansi_state, STATE_QUOTE_LINE);
+            STATE_SET(ansi_state, STATE_NEW_LINE);
+        } else {
+            if (STATE_ISSET(ansi_state, STATE_ESC_SET)) {
+                if (buf[i] == 'm') {
+                    /*
+                     *[0;1;4;31m */
+                    if (STATE_ISSET(ansi_state, STATE_FONT_SET)) {
+                        BUFFERED_OUTPUT(output, "</font>", 7);
+                        STATE_CLR(ansi_state, STATE_FONT_SET);
+                    }
+                    if (i < buflen - 1) {
+                        generate_font_style(&font_style, ansi_val, ival + 1);
+                        if (STATE_ISSET(ansi_state, STATE_QUOTE_LINE))
+                            STYLE_SET(font_style, FONT_STYLE_QUOTE);
+                        print_font_style(font_style, output);
+                        STATE_SET(ansi_state, STATE_FONT_SET);
+                        STATE_CLR(ansi_state, STATE_ESC_SET);
+                        /*
+                         * STYLE_ZERO(font_style);
+                         */
+                        /*
+                         * clear ansi_val[] array
+                         */
+                        bzero(ansi_val, sizeof(ansi_val));
+                        ival = 0;
+                    }
+                } else if (isalpha(buf[i])) {
+                    /*
+                     *[23;32H */
+                    /*
+                     * ignore it
+                     */
+                    STATE_CLR(ansi_state, STATE_ESC_SET);
+                    STYLE_ZERO(font_style);
+                    /*
+                     * clear ansi_val[] array
+                     */
+                    bzero(ansi_val, sizeof(ansi_val));
+                    ival = 0;
+                    continue;
+                } else if (buf[i] == ';') {
+                    if (ival < sizeof(ansi_val) - 1) {
+                        ival++; /* go to next ansi_val[] element */
+                        ansi_val[ival] = 0;
+                    }
+                } else if (buf[i] >= '0' && buf[i] <= '9') {
+                    ansi_val[ival] *= 10;
+                    ansi_val[ival] += (buf[i] - '0');
+                } else {
+                    /*
+                     *[1;32/XXXX or *[* or *[[ */
+                    /*
+                     * not a valid ANSI string, just output it
+                     */
+                    size_t len;
+
+                    ansi_end = &buf[i];
+                    len = ansi_end - ansi_begin + 1;
+                    print_raw_ansi(ansi_begin, len, output);
+                    STATE_CLR(ansi_state, STATE_ESC_SET);
+                    /*
+                     * clear ansi_val[] array
+                     */
+                    bzero(ansi_val, sizeof(ansi_val));
+                    ival = 0;
+                }
+
+            } else
+                print_raw_ansi(&buf[i], 1, output);
+        }
+	}
+    if (STATE_ISSET(ansi_state, STATE_FONT_SET)) {
+        BUFFERED_OUTPUT(output, "</font>", 7);
+        STATE_CLR(ansi_state, STATE_FONT_SET);
+    }
+	BUFFERED_FLUSH(output);
+}
+#endif
+
 /**
  * Warning: Use of this function is deprecated. It's kept only for compatible
  * purpose. Use output_ansi_text() instead.
Index: src/maintain.c
===================================================================
--- src/maintain.c	(revision 10670)
+++ src/maintain.c	(working copy)
@@ -794,7 +794,7 @@
 }
 int modify_board(int bid)
 {
-#define MB_ITEMS 24
+#define MB_ITEMS 25
     FILE *fp;
     struct _select_item sel[MB_ITEMS+1];
     struct _select_def conf;
@@ -810,7 +810,7 @@
         "[6]转信标签  :","[7]讨论区描述:","[8]匿名讨论区:","[9]统计文章数:","[A]统计十大  :",
         "[B]目录讨论区:","[C]所属目录  :","[D]向外转信  :","[E]上传附件  :","[F]E-mail发文:",
         "[G]不可回复  :","[H]读限制Club:","[I]写限制Club:","[J]隐藏Club  :","[K]精华区位置:",
-        "[L]权限限制  :","[M]身份限制  :","[N]积分限制  :","[Q][退出]    :"
+        "[L]权限限制  :","[M]身份限制  :","[N]积分限制  :","[P]强制模板  :","[Q][退出]    :"
     };
     change=0; loop=1;
     /*选择讨论区*/
@@ -844,9 +844,12 @@
         } else if (i==22) {
             sel[i].x=42;
             sel[i].y=17;
+        } else if (i==23) {
+            sel[i].x=2;
+            sel[i].y=18;
         } else if (i==MB_ITEMS-1) {
             sel[i].x=2;
-            sel[i].y=18;
+            sel[i].y=19;
         } else {
             sel[i].x=2;
             sel[i].y=i-4;
@@ -947,6 +950,10 @@
     sprintf(menustr[22],"%-15s%s <%d>",menuldr[22],"无效选项",bh.score_level);
 #endif /* HAVE_USERSCORE */
     /*退出*/
+    /*强制模板发文*/
+    sel[23].hotkey='P';
+    sprintf(menustr[23],"%-15s%s",menuldr[23],(bh.flag&BOARD_TMP_POST)?"是":"否");
+    /*退出*/
     sel[MB_ITEMS-1].hotkey='Q';
     sprintf(menustr[MB_ITEMS-1],"%-15s%s",menuldr[MB_ITEMS-1],change?"\033[1;31m已修改\033[m":"未修改");
     sel[MB_ITEMS].x=-1; sel[MB_ITEMS].y=-1; sel[MB_ITEMS].type=0; sel[MB_ITEMS].hotkey=-1; sel[MB_ITEMS].data=NULL;
@@ -1599,6 +1606,19 @@
 #endif /* HAVE_USERSCORE */
                 break;
                 /*退出*/
+            /* 强制模板发文 */
+            case 23:
+                newbh.flag^=BOARD_TMP_POST;
+                /*标记修改状态*/
+                if ((bh.flag&BOARD_TMP_POST)^(newbh.flag&BOARD_TMP_POST)) {
+                    sprintf(menustr[23],"%-15s\033[1;32m%s\033[m",menuldr[23],(newbh.flag&BOARD_TMP_POST)?"是":"否");
+                    change|=(1<<23);
+                } else {
+                    sprintf(menustr[23],"%s",orig[23]);
+                    change&=~(1<<23);
+                }
+                break;
+                /*退出*/
             case MB_ITEMS-1:
                 if (change) {
                     /*冲突检测及确认*/
Index: src/bbs.c
===================================================================
--- src/bbs.c	(revision 10670)
+++ src/bbs.c	(working copy)
@@ -3375,7 +3375,7 @@
     if (!anonyboard)
 #endif
         modify_user_mode(POSTING);
-
+	if (bp->flag&BOARD_TMP_POST) use_tmpl = 1;
     if (use_tmpl > 0) {
         FILE *fp,*fp1;
         char filepath1[STRLEN];
Index: src/bbs.h
===================================================================
--- src/bbs.h	(revision 10670)
+++ src/bbs.h	(working copy)
@@ -312,6 +312,7 @@
 #define BOARD_POSTSTAT      0x1000  /* 不统计十大 */
 #define BOARD_NOREPLY       0x2000  /* 不可 re 文 */
 #define BOARD_RULES         0x4000  /* 治版方针 ok */
+#define BOARD_TMP_POST		0x8000	/* 强制模板发文 */ /*Alex@bupt add on July 9th, 2010*/
 /* boardheader.flag 的最高八位留给用户自定义用途: 0xXX000000 */
 
 #define ZAPPED              0x1     /* For boards...tells if board is Zapped */
Index: src/announce.c
===================================================================
--- src/announce.c	(revision 10670)
+++ src/announce.c	(working copy)
@@ -1867,9 +1867,6 @@
                         }
                         ret = post_cross(getCurrentUser(),bh,"",M_ITEM(&me, me.now)->title,fname,0,false,ans[0],5,getSession());
                         switch (ret) {
-                            case 1 :
-                                prints("\033[1;32m%s\033[0;33m<Enter>\033[m","转载完成!");
-                                break;
                             case -1:
                                 prints("\033[1;33m%s\033[0;33m<Enter>\033[m", "转载过程中发生错误 ...");
                                 break;
@@ -1881,6 +1878,8 @@
                                 move(t_lines - 1, 0);
                                 prints("%s", "                              \x1b[33m请按 ◆\x1b[36mEnter\x1b[33m◆ 继续\x1b[m");
                                 break;
+                            default:
+                                prints("\033[1;32m%s\033[0;33m<Enter>\033[m","转载完成!");
                         }
                         WAIT_RETURN;
                     } while (0);
Index: libBBS/article.c
===================================================================
--- libBBS/article.c	(revision 10670)
+++ libBBS/article.c	(working copy)
@@ -1144,7 +1144,7 @@
 #endif /* SMTH */
     if (after_post(user, &postfile, toboard->filename, NULL, !(Anony), session) == -2)
         return -2;
-    return 1;
+    return postfile.id;
 }
 
 
@@ -1159,8 +1159,7 @@
     if (getbid(nboard, &toboard) <= 0) {       /* 搜索要POST的版 ,判断是否存在该版 */
         return -1;
     }
-    post_cross(user, toboard, fromboard, posttitle, filename, Anony, false, 'l', mode, session);  /* post 文件 */
-    return 0;
+    return post_cross(user, toboard, fromboard, posttitle, filename, Anony, false, 'l', mode, session);  /* post 文件 */
 }
 
 
